### ISSUE TITLE
Light Controls "jumps" on iOS but not on macOS 

### ISSUE THREAD

**Issue by odx** [arachnetech/homebridge-mqttthing#122]
Since a few weeks I have a strange behavior of my IKEA Tradfi Lamps connected through zigbee2mqtt:
Whenever is adjust the settings (brightness or color) for the lamps the setting Strats jumping back and forth. For color it seems that the lamp continuously jumps between the old and the new setting in an endless loop. For brightness it changes between several values (I assume intermediate brightness values that were set when adjusting the brightness slider).

When I use automations to set the values for brightness or color it works just fine. Also when I use the Mac app "Home" instead of the iOS App I can set the values just fine. The Mac App does not update the slider values during movement so I assume it only sets the characteristic once I release the mouse button which prevents this behavior.

I also tried setting the values with the eglato Eve App which transmits the values also "live". It seems like the transmission of the values is a lot slower but the app manages to adjust the values in the correct order so it just "lags" a bit but it does not get stuck in an endless loop.

Yes I know it could be an issue with the apple home app, but I wonder if there is anything mqttthing can do to compensate this since apple will probably not accept a bug report for uncertified HomeKit solutions...

**Comment by arachnetech** [comment]
Very strange behaviour. Have you tried subscribing to the MQTT messages being published by mqttthing to see what it's sending?

Do you have different set and get topics? If not, I can imagine that causing problems like this.

**Comment by odx** [comment]
Yes, I subscribed to the topic and it reflects exactly what is shown in the Home App and what the lamp does. Switching back and forth between the settings.

This is my configuration. Maybe it is related to the fact that the topics for On and HSV are the same, just using different payload?

```
{
            "accessory": "mqttthing",
            "username": "<username>",
            "password": "<password>",
            "type": "lightbulb",
            "name": "Sofalampe",
            "topics": {
                "getOn": {
                    "topic": "zigbee/sofalampe",
                    "apply": "return JSON.parse(message).state;"
                },
                "setOn": {
                    "topic": "zigbee/sofalampe/set",
                    "apply": "return JSON.stringify({ state: message})"
                },
                "getHSV": {
                    "topic": "zigbee/sofalampe",
                    "apply": "function cie_to_rgb(r,s,t){void 0===t&&(t=254);var a=1-r-s,h=(t/254).toFixed(2),o=h/s*r,e=h/s*a,v=1.656492*o-.354851*h-.255038*e,i=.707196*-o+1.655397*h+.036152*e,n=.051713*o-.121364*h+1.01153*e;return v>n&&v>i&&v>1?(i/=v,n/=v,v=1):i>n&&i>v&&i>1?(v/=i,n/=i,i=1):n>v&&n>i&&n>1&&(v/=n,i/=n,n=1),v=v<=.0031308?12.92*v:1.055*Math.pow(v,1/2.4)-.055,i=i<=.0031308?12.92*i:1.055*Math.pow(i,1/2.4)-.055,n=n<=.0031308?12.92*n:1.055*Math.pow(n,1/2.4)-.055,v=Math.round(255*v),i=Math.round(255*i),n=Math.round(255*n),isNaN(v)&&(v=0),isNaN(i)&&(i=0),isNaN(n)&&(n=0),[v,i,n]}function RGBtoHSV(r,s,t){1===arguments.length&&(s=r.g,t=r.b,r=r.r);var a,h=Math.max(r,s,t),o=Math.min(r,s,t),e=h-o,v=0===h?0:e/h,i=h/255;switch(h){case o:a=0;break;case r:a=s-t+e*(s<t?6:0),a/=6*e;break;case s:a=t-r+2*e,a/=6*e;break;case t:a=r-s+4*e,a/=6*e}return{h:a,s:v,v:i}}var m=JSON.parse(message),rgb=cie_to_rgb(m.color.x,m.color.y,m.brightness/2.55),hsv=RGBtoHSV(rgb[0],rgb[1],rgb[2]);hsv.h=360*hsv.h,hsv.s=100*hsv.s,hsv.v=254*hsv.v;var result=hsv.h.toString()+\",\"+hsv.s.toString()+\",\"+(m.brightness/2.54).toString();return result;"
                },
                "setHSV": {
                    "topic": "zigbee/sofalampe/set",
                    "apply": "function HSVtoRGB(r,a,e){var t,s,o,b,c,n,i,u;switch(1===arguments.length&&(a=r.s,e=r.v,r=r.h),n=e*(1-a),i=e*(1-(c=6*r-(b=Math.floor(6*r)))*a),u=e*(1-(1-c)*a),b%6){case 0:t=e,s=u,o=n;break;case 1:t=i,s=e,o=n;break;case 2:t=n,s=e,o=u;break;case 3:t=n,s=i,o=e;break;case 4:t=u,s=n,o=e;break;case 5:t=e,s=n,o=i}return{r:Math.round(255*t),g:Math.round(255*s),b:Math.round(255*o)}}function ScaledHSVtoRGB(r,a,e){return HSVtoRGB(r/360,a/100,e/100)}function rgb_to_cie(r,a,e){var t=.664511*(r=r>.04045?Math.pow((r+.055)/1.055,2.4):r/12.92)+.154324*(a=a>.04045?Math.pow((a+.055)/1.055,2.4):a/12.92)+.162028*(e=e>.04045?Math.pow((e+.055)/1.055,2.4):e/12.92),s=.283881*r+.668433*a+.047685*e,o=88e-6*r+.07231*a+.986039*e,b=(t/(t+s+o)).toFixed(4),c=(s/(t+s+o)).toFixed(4);return isNaN(b)&&(b=0),isNaN(c)&&(c=0),[b,c]}var params=message.split(\",\"),result={},rgb=ScaledHSVtoRGB(params[0],params[1],100),xy=rgb_to_cie(rgb.r,rgb.g,rgb.b);result.color={x:xy[0],y:xy[1]};var b=2.54*params[2];result.brightness=b;return JSON.stringify(result);"
                }
            },
            "onValue": "ON",
            "offValue": "OFF"
        }
```

**Comment by czhDavid** [comment]
Might be helpfull when i subscribe to the MQTT topic with paho python client i get 4x message received. I think that is where the problem lies

**Comment by arachnetech** [comment]
Not sure whether it helps, but there are new setRGB and getRGB topics available now which might simplify your configuration. If I remember correctly, if you configure setRGB but not setOn then ‘off’ is represented as RGB(0,0,0).

**Comment by EjvindHald** [comment]
Look at [this](https://community.openhab.org/t/announcing-some-improvements-to-the-homekit-plugin-for-openhab2/62729/118) with a post from Frank Steiler 18 June 2019, where he explains how he solved this. The same odd behaviour is seen in the native Homekit binding for openHAB, so it is probably a weakness in the Apple homeapp as also suggested by @odx 
Looping in @arachnetech because it might be better to solve it in this plug-in rather than all the mqtt clients.

**Comment by AndrzejOlender** [comment]
I have the same problem as the author. Ikea RGB bulb, if I remove getHSV it works fine, but I lose a lot.

{  
        "accessory":"mqttthing",
        "type":"lightbulb",
        "name":"Ikea Lamp 3",
        "topics": {  
           "getOn": {  
              "topic": "zigbee2mqtt/IkeaBulbLamp",
              "apply": "return JSON.parse(message).state;"
           },
           "setOn": {  
              "topic": "zigbee2mqtt/IkeaBulbLamp/set",
              "apply": "return JSON.stringify({ state: message})"
           },
           "getHSV": {
            "topic": "zigbee2mqtt/IkeaBulbLamp",
            "apply": "function cie_to_rgb(r,s,t){void 0===t&&(t=254);var a=1-r-s,h=(t\/254).toFixed(2),o=h\/s*r,e=h\/s*a,v=1.656492*o-.354851*h-.255038*e,i=.707196*-o+1.655397*h+.036152*e,n=.051713*o-.121364*h+1.01153*e;return v>n&&v>i&&v>1?(i\/=v,n\/=v,v=1):i>n&&i>v&&i>1?(v\/=i,n\/=i,i=1):n>v&&n>i&&n>1&&(v\/=n,i\/=n,n=1),v=v<=.0031308?12.92*v:1.055*Math.pow(v,1\/2.4)-.055,i=i<=.0031308?12.92*i:1.055*Math.pow(i,1\/2.4)-.055,n=n<=.0031308?12.92*n:1.055*Math.pow(n,1\/2.4)-.055,v=Math.round(255*v),i=Math.round(255*i),n=Math.round(255*n),isNaN(v)&&(v=0),isNaN(i)&&(i=0),isNaN(n)&&(n=0),[v,i,n]}function RGBtoHSV(r,s,t){1===arguments.length&&(s=r.g,t=r.b,r=r.r);var a,h=Math.max(r,s,t),o=Math.min(r,s,t),e=h-o,v=0===h?0:e\/h,i=h\/255;switch(h){case o:a=0;break;case r:a=s-t+e*(s<t?6:0),a\/=6*e;break;case s:a=t-r+2*e,a\/=6*e;break;case t:a=r-s+4*e,a\/=6*e}return{h:a,s:v,v:i}}var m=JSON.parse(message),rgb=cie_to_rgb(m.color.x,m.color.y,m.brightness\/2.55),hsv=RGBtoHSV(rgb[0],rgb[1],rgb[2]);hsv.h=360*hsv.h,hsv.s=100*hsv.s,hsv.v=254*hsv.v;var result=hsv.h.toString()+\",\"+hsv.s.toString()+\",\"+(m.brightness\/2.54).toString();return result;"
          },
           "setHSV": {  
              "topic": "zigbee2mqtt/IkeaBulbLamp/set",
              "apply": "function HSVtoRGB(r,a,e){var t,s,o,b,c,n,i,u;switch(1===arguments.length&&(a=r.s,e=r.v,r=r.h),n=e*(1-a),i=e*(1-(c=6*r-(b=Math.floor(6*r)))*a),u=e*(1-(1-c)*a),b%6){case 0:t=e,s=u,o=n;break;case 1:t=i,s=e,o=n;break;case 2:t=n,s=e,o=u;break;case 3:t=n,s=i,o=e;break;case 4:t=u,s=n,o=e;break;case 5:t=e,s=n,o=i}return{r:Math.round(255*t),g:Math.round(255*s),b:Math.round(255*o)}}function ScaledHSVtoRGB(r,a,e){return HSVtoRGB(r\/360,a\/100,e\/100)}function rgb_to_cie(r,a,e){var t=.664511*(r=r>.04045?Math.pow((r+.055)\/1.055,2.4):r\/12.92)+.154324*(a=a>.04045?Math.pow((a+.055)\/1.055,2.4):a\/12.92)+.162028*(e=e>.04045?Math.pow((e+.055)\/1.055,2.4):e\/12.92),s=.283881*r+.668433*a+.047685*e,o=88e-6*r+.07231*a+.986039*e,b=(t\/(t+s+o)).toFixed(4),c=(s\/(t+s+o)).toFixed(4);return isNaN(b)&&(b=0),isNaN(c)&&(c=0),[b,c]}var params=message.split(\",\"),result={},rgb=ScaledHSVtoRGB(params[0],params[1],100),xy=rgb_to_cie(rgb.r,rgb.g,rgb.b);result.color={x:xy[0],y:xy[1]};var b=2.54*params[2];result.brightness=b;return JSON.stringify(result);"
           }
        },
        "onValue":"ON",
        "offValue":"OFF"
}

**Comment by EjvindHald** [comment]
I think that for dimmer code similar to Frank Steiler needs to be implemented with at delay on 125 ms. 
It is described at this [link](https://community.openhab.org/t/announcing-some-improvements-to-the-homekit-plugin-for-openhab2/62729/118)  18 June 2019, and the actual code can be viewed [here](https://github.com/steilerDev/homebridge-openhab2-complete/blob/master/accessory/characteristic/SetAndCommitLight.js) and [here](https://github.com/steilerDev/homebridge-openhab2-complete/blob/master/accessory/characteristic/SetAndCommit.js)
Unfortunately, I do not know how to do this..

**Comment by odx** [comment]
@EjvindHald after your first reply to this issue I implemented a rough test to delay the messages received from HomeKit (and only fire the last message after a given timeout without change has been reached) using [debounce](https://www.npmjs.com/package/debounce) with no luck. Everything that happened was that the jumpy behavior slowed down but did not get resolved.

My approach would be to have a closer look at the order and senders of the MQTT messages but I still did not find the time to analyze this in depth.

**Comment by arachnetech** [comment]
I will try to look at this properly - very busy at work at the moment, unfortunately!

It would be interesting to see MQTT capture from MacOS compared to iOS to compare the messages sent, as this might indicate the source of the flicker.

Perhaps I am sending duplicate ‘on’ messages which alternate with the HSV ones and cause the flicker? I will try looking at MQTT output with your config myself when I get a chance...

**Comment by AndrzejOlender** [comment]
@arachnetech Thanks for your commitment.

**Comment by EjvindHald** [comment]
@arachnetech : I suspect that you do eveything correct and no duplicate messages. In the native Homekit add on to openHAB there is exactly same problem.

In the coming weekend I will try the same dimmer hardware configured in openHAB to go through homebridge using both mqtt and openhab complete homebridge plugin. If it works in one but not the other, we are step closer to root cause.

**Comment by EjvindHald** [comment]
I have now tried with the same dimmer hardware in plug-ins homebridge-mqqthing and homebridge-openhab2-complete. Both of these are working correctly with no "jumps" on iOS. I have verified that the behaviour is identical in IOS version 12 and 13.

**Comment by czhDavid** [comment]
I think you got lucky on the Mac OS. Anyway I tried to implement it myself from scratch and found out it has nothing to do with homebridge. It is actually problem with zigbee2mqtt. For more information you can look into this[zigbee2mqtt issue](https://github.com/Koenkk/zigbee2mqtt/issues/1722). Although not stated directly in release notes for version 1.6 the issue of duplicate messages should be resolved! I have ma CC debuger in office so I might try it tomorow. Will let you know the result

**Comment by odx** [comment]
I just retested this with the iOS 13.1.1 and I could not reproduce this anymore. Although it was still present in 13.0 it seems like apple did some rewrite (as they did with many other apps in iOS 13) which seems to resolve this issue.

**Comment by VOVAiS** [comment]
if you delete topic setON, the brightness adjustment will work correctly. when you adjust the brightness every time the topic is sent to SetON. for this reason the light jumps

**Comment by arachnetech** [comment]
Reoopening as this has potentially resurfaced in #322.

**Comment by arachnetech** [comment]
Debounce option implemented in v1.1.25.
